Dev Container Support

sora-mono opened the pull request:
Currently I only finished linux Dockerfile and launch scripts

As for the plan, there is still a lot work to do:
1. Support Linux and Windows dev-container
2. Support Docker image build CI/CD, with cache to acclerate build speed
3. Move all package version and hash away from cmake and xmake to separated config folder, which is easy for share and manage between different systems
4. Support toolchain version config
5. Modify CI to build everything with docker
6. Test all build function could work in docker image

How to try this preview version:
cd clice # clice project root dir
docker/linux/build.sh
docker/linux/run.sh

miyanyan commented:
you can add uv to the dev container so the users can run tests

aurora0x27 commented:
There is a space in front of the `.dockerignore` file, is this intentional?
<img width="678" height="606" alt="图片" src="https://github.com/user-attachments/assets/1e3b917f-74cf-468b-849b-e8fee00a8833" />

sora-mono commented:
> you can add uv to the dev container so the users can run tests

Thank you for your advice, I will convert from apt to uv for python and python library management

miyanyan commented:
> > you can add uv to the dev container so the users can run tests
> 
> @miyanyan Thank you for your advice, I will convert from apt to uv for python and python library management

Yeah, they use uv for test https://github.com/clice-io/clice/blob/2ffab4fdffeb29cefe72bb9d16731e2c05840108/.github/workflows/cmake.yml#L94

16bit-ykiko commented:
There is an extra space in dockignore filename https://github.com/clice-io/clice/blob/main/%20.dockerignore, could you also fix it?

sora-mono commented:
> There is an extra space in dockignore filename https://github.com/clice-io/clice/blob/main/%20.dockerignore, could you also fix it?

Sure, I'll fix it together

sora-mono commented:
Update:
* python management changed from apt to uv
* Parallel dependency install and download support
* docker buildx cache for llvm-binary download
* bug fix

sora-mono commented:
Update:

1. XMake package cache support in Dockerfile
2. Install exactly only one toolchain
3. Fixed async crash in dockerfile
4. Refactored package install sequence and source for build speed and avoid other compile tool chain installation

clang toolchain and xmake cache are ready
However lld would report unresolved symbols, which is still on inspecting

star-hengxing commented on xmake.lua:
```diff
@@ -42,7 +42,7 @@ end
 add_defines("TOML_EXCEPTIONS=0")
 add_requires(libuv_require, "spdlog[header_only=n,std_format,noexcept]" ,"toml++")
-add_requires("llvm", {system = false})
+add_requires("clice-llvm", {system = false})
```
You can use `add_requires("clice-llvm", {alias = "llvm"})` && `add_packages("llvm")`

16bit-ykiko commented:
/gemini summary

miyanyan commented:
There are two docker files, the Dockerfile(dev, linux + windows finally) in this pr, and https://github.com/clice-io/clice/blob/main/Dockerfile for release, is maintaining the two docker files the right way? It seems they have the same points, only the docker stage they are in is different.

16bit-ykiko commented:
Remove this file https://github.com/clice-io/clice/blob/main/Dockerfile?

sora-mono commented:
@16bit-ykiko @miyanyan I plan to replace all builds including CI with dev-container, but not now.
Also, this PR only contains Linux platform, and it will be refactored to support static linking libstdc++.
Althouth not completely supported, I decided to merge only Linux because someone requested for this feature merged as quickly as possible, it could be very annoying for them to prepare toolchain in Linux, so this PR would be a PREVIEW feature and not complete resolution.
As for Windows platform toolchain, it is easy to install and use(I fixed msvc build in another PR and those who build on windows could build with the latest msvc toolchain, without the need to install clang)
So the Dockerfile in root should not be removed immediately, there's already a github workflow for it, and could be removed until this PR is completely supported.
